packages/zoe/src/zoeService/zoe.js

```diff
@@ -0,0 +1,357 @@
+        .done()
+        .then(
+          () => exitAllSeats(),
+          () => exitAllSeats(),
+        );
+
+      const registerIssuerByKeyword = (keyword, issuer, brand) => {
+        instanceRecord.issuerKeywordRecord = {
+          ...instanceRecord.issuerKeywordRecord,
+          [keyword]: issuer,
+        };
+        instanceRecord.brandKeywordRecord = {
+          ...instanceRecord.brandKeywordRecord,
+          [keyword]: brand,
+        };
+      };
+
+      /** @type MakeZoeMint */
+      const makeZoeMint = (keyword, mathHelperName = 'nat') => {
+        // Local indicates one that zoe itself makes from vetted code,
+        // and so can be assumed correct and fresh by zoe.
+        const {
+          mint: localMint,
+          issuer: localIssuer,
+          amountMath: localAmountMath,
+          brand: localBrand,
+        } = makeIssuerKit(keyword, mathHelperName);
+        const localIssuerRecord = harden({
+          brand: localBrand,
+          issuer: localIssuer,
+          amountMath: localAmountMath,
+        });
+        issuerTable.registerIssuerRecord(localIssuerRecord);
+        registerIssuerByKeyword(keyword, localIssuer, localBrand);
+        const localPooledPurse = localIssuer.makeEmptyPurse();
+        brandToPurse.init(localBrand, localPooledPurse);
+
+        /** @type ZoeMint */
+        const zoeMint = harden({
+          getIssuerRecord: () => {
+            return localIssuerRecord;
+          },
+          mintAndEscrow: totalToMint => {
+            const payment = localMint.mintPayment(totalToMint);
+            localPooledPurse.deposit(payment, totalToMint);
+          },
+          withdrawAndBurn: totalToBurn => {
+            const payment = localPooledPurse.withdraw(totalToBurn);
+            localIssuer.burn(payment, totalToBurn);
+          },
+        });
+        return zoeMint;
+      };
+
+      const bundle = installation.getBundle();
+      const addSeatObjPromiseKit = makePromiseKit();
+      const publicFacetPromiseKit = makePromiseKit();
+
+      /** @type {InstanceAdmin} */
+      const instanceAdmin = {
+        addZoeSeatAdmin: async (invitationHandle, zoeSeatAdmin, seatData) => {
+          zoeSeatAdmins.add(zoeSeatAdmin);
+          return E(
+            /** @type Promise<addSeatObj> */ (addSeatObjPromiseKit.promise),
+          ).addSeat(invitationHandle, zoeSeatAdmin, seatData);
+        },
+        hasZoeSeatAdmin: zoeSeatAdmin => zoeSeatAdmins.has(zoeSeatAdmin),
+        removeZoeSeatAdmin: zoeSeatAdmin => zoeSeatAdmins.delete(zoeSeatAdmin),
+        getPublicFacet: () => publicFacetPromiseKit.promise,
+        getTerms: () => instanceRecord.terms,
+        getIssuers: () => instanceRecord.issuerKeywordRecord,
+        getBrands: () => instanceRecord.brandKeywordRecord,
+        getInstance: () => instance,
+      };
+
+      instanceToInstanceAdmin.init(instance, instanceAdmin);
+
+      /** @type {ZoeInstanceAdmin} */
+      const zoeInstanceAdminForZcf = {
```

Comment:
There's an interesting failure property that we're not setup to handle well. Generally, if our caller gives us bad arguments, we're doing our job if we detect these bad argument during input validation, perform no side effects, and throw an error or reject a promise. However, if our zcf client gives us something bad, that's a "should never happen", and zoe should treat this as an immediately fatal error of that zcf. We should terminate it or treat it as terminated, rather than just reporting an error to it.

I'm not sure how to factor this out so we get this effect without repeated boilerplate and maintenance headache.

Reply:
https://github.com/Agoric/agoric-sdk/issues/1500